### ISSUE TITLE
PTQ INT8 via TFLiteConverter — encoder-decoder seq2seq model loses encoder context entirely after conversion

### ISSUE THREAD

**Issue by AminLO-ma** [tensorflow/tensorflow#117052]
### 1. System information

- **OS Platform and Distribution:** macOS (Apple M2 Pro, 32GB) — conversion
- **TensorFlow installation:** pip package
- **TensorFlow version:** 2.13
- **Python version:** 3.11.14
- **transformers version:** 4.40
- **Target inference platform:** embedded Linux, INT8/INT16 hardware delegate only

---

#### Option B — Conversion code

```python
import numpy as np
import tensorflow as tf
from transformers import MarianTokenizer, TFMarianMTModel
from datasets import load_dataset

# 1. Load model
model_name = "Helsinki-NLP/opus-mt-en-fr"
tok   = MarianTokenizer.from_pretrained(model_name)
model = TFMarianMTModel.from_pretrained(model_name)

MAX_LEN = 64

# 2. Calibration dataset (200 samples from opus_books EN-FR)
ds = load_dataset("opus_books", "en-fr", split="train", streaming=True).take(200)

def prepare_calibration_data():
    calibration_data = []
    for example in ds:
        enc = tok(example['translation']['en'], max_length=MAX_LEN,
                  padding="max_length", truncation=True, return_tensors="np")
        dec = tok(text_target=example['translation']['fr'], max_length=MAX_LEN,
                  padding="max_length", truncation=True, return_tensors="np")
        calibration_data.append({
            "enc_ids":  enc['input_ids'].astype(np.int32),
            "enc_mask": enc['attention_mask'].astype(np.int32),
            "dec_ids":  dec['input_ids'].astype(np.int32),
            "dec_mask": dec['attention_mask'].astype(np.int32),
        })
    return calibration_data

calib_samples = prepare_calibration_data()

def representative_dataset():
    for sample in calib_samples:
        yield [sample["enc_ids"], sample["enc_mask"],
               sample["dec_ids"], sample["dec_mask"]]

# 3. Fused encoder-decoder concrete function
@tf.function(input_signature=[
    tf.TensorSpec(shape=[1, MAX_LEN], dtype=tf.int32, name="encoder_input_ids"),
    tf.TensorSpec(shape=[1, MAX_LEN], dtype=tf.int32, name="encoder_attention_mask"),
    tf.TensorSpec(shape=[1, MAX_LEN], dtype=tf.int32, name="decoder_input_ids"),
    tf.TensorSpec(shape=[1, MAX_LEN], dtype=tf.int32, name="decoder_attention_mask"),
])
def fused_model(encoder_input_ids, encoder_attention_mask,
                decoder_input_ids, decoder_attention_mask):
    return model(
        input_ids=encoder_input_ids,
        attention_mask=encoder_attention_mask,
        decoder_input_ids=decoder_input_ids,
        decoder_attention_mask=decoder_attention_mask,
        training=False
    ).logits

# 4. PTQ INT8 conversion
converter = tf.lite.TFLiteConverter.from_concrete_functions(
    [fused_model.get_concrete_function()], fused_model
)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_dataset
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_output_type = tf.float32

tflite_model = converter.convert()

with open("opus_mt_en_fr_int8.tflite", "wb") as f:
    f.write(tflite_model)
```

**Conversion warning noted:**

Statistics for quantized inputs were expected, but not specified; continuing anyway.
Non-Converted Ops: 274 / 1072 (25.56%) — 274 ARITH ops (f32 constants)

---

### 3. Failure after conversion

✅ Conversion succeeds without errors.
❌ **Model produces wrong results — severely degraded accuracy.**
 
` FP32 BLEU : 23.90
INT8 BLEU : 0.04 → -99.9% `

The model generates repeated tokens regardless of input (e.g. `l l l l...`). The decoder stops using encoder context entirely from the first inference step — confirmed by comparing FP32 vs INT8 logits position by position: INT8 returns the same top-5 tokens at every position regardless of source input.

Root cause hypothesis: Q (decoder-side) and K/V (encoder-side) cross-attention projections are assigned mismatched per-tensor quantization scales by the converter, corrupting the `QK^T` dot product.

Note: `EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8` is not viable — `TILE` op unsupported at runtime.

---

### 5. Any other info / logs

` fully_quantize: 0, inference_type: 6,
input_inference_type: FLOAT32, output_inference_type: FLOAT32
Estimated count of arithmetic ops: 5.792 G ops, equivalently 2.896 G MACs`

Reproducible notebook available on request.

Open to any working approach to move forward.

**Comment by aravindmurali22** [comment]
@sourcelite, could you please look into this? Thank you.

**Comment by shivvaam0001** [comment]
Hi @AminLO-ma,

Thank you for taking the time to report this and for your patience.
TFLite has transitioned to LiteRT, Google's primary runtime for on-device AI. All active development, feature updates, and performance enhancements have shifted exclusively to LiteRT.
We encourage you to try reproducing the issue using LiteRT. If you still encounter this issue, please open a new issue with the updated details in the [LiteRT Repository](https://github.com/google-ai-edge/LiteRT).
For more information and step-by-step guidance on updating your project, please refer to the [LiteRT Migration Guide](https://developers.google.com/edge/litert/migration).

Thank you again for your support and understanding!
